### ISSUE TITLE
Integrating 10 visium Hd samples

### ISSUE THREAD

**Issue by raygozag** [satijalab/seurat#9374]
Hi i'm trying to integrate 10 Visium HD samples, and there is an issue when merging the last sample overflowing 2^32 elements. I'm trying to make work some other answer, but my question is: can I integrate the first N samples and then and then integrate the result of this with the next sample and so on iteratively?

is this equivalent to integrating all ten samples at once?

**Comment by wangbenwang123** [comment]
I’m not sure about the exact issue you're encountering. Could you please share the specific code and error message you're seeing?

**Comment by raygozag** [comment]
Well my question wasn't answered.

The other issue is:

it takes a long time to reproduce but it's the same issue as in the posts below in IntegrateData on seurat v5

This happens on the 10th sample merge, and it's on TransformMatrix and there's supposed to be a fork of seurat with a fix but is for 4.3.

https://github.com/satijalab/seurat/issues/8963
https://github.com/satijalab/seurat/discussions/7311

https://stackoverflow.com/questions/54405435/sparse-matrix-support-for-long-vectors-over-231-elements

**Comment by raygozag** [comment]
Well my question is: can I integrate the first N samples and then and then integrate the result of this with the next sample and so on iteratively?

is this equivalent to integrating all ten samples at once?

**Comment by mccormickceci** [comment]
Hi,
Would you be able to provide some more context on your question: are you attempting to use merge() or are you using IntegrateLayers() in order to combine your samples?

**Comment by yiqun0423** [comment]
Hi, 
In response to your question, I encountered a similar issue regarding the integration of 3 Visium HD samples. In each 6.5x 6.5 capture area, I have one sample in it. After selecting the region of interest based on HE morphology, I ran an automatic spatial clustering, which generated different numbers of clusters for each sample. I wonder how to normalise the Visium HD data among samples when doing the clustering, PCA analysis, annotation, etc. Is there any standard pipeline regarding Visium HD data integration? Biologically, it is important to integrate 3 samples within one group.

**Comment by rsatija** [comment]
there is no uniform single answer to this question. it is certainly possible to integrate a subset of datasets, and then map the remainder to the integrated reference (as we do here https://satijalab.org/seurat/articles/integration_mapping). this is not equivalent to integrating everything together, but is often justified (for example, if a subset of the samples are comprehensive and high-quality to define a reference)

**Comment by brianlamere** [comment]
> there is no uniform single answer to this question. it is certainly possible to integrate a subset of datasets, and then map the remainder to the integrated reference (as we do here https://satijalab.org/seurat/articles/integration_mapping). this is not equivalent to integrating everything together, but is often justified (for example, if a subset of the samples are comprehensive and high-quality to define a reference)

The issue they likely hit, or at least the issue I'm hitting that seems related, results in this error:

`p[length(p)] cannot exceed 2^31-1`

It was briefly discussed at https://github.com/satijalab/seurat/issues/8963 - and tonight I'll be trying to figure out how to "integrate a subset of datasets, and then map the remainder to the integrated reference" - though I'm hoping it's not too discordant "to integrating everything together" since there is a seurat code limitation we're hitting :)

**Comment by raygozag** [comment]
yes this was one of the issues, among inefficient code.  but i was recently able to integrate 20 visium hd samples.

**Comment by brianlamere** [comment]
> yes this was one of the issues, among inefficient code. but i was recently able to integrate 20 visium hd samples.

well this is 12 samples of 10x single cell multiome, a comment on the subject on the other issue I mentioned that previously the task died when it used up 384G addressable space (128ram, 256 swap), so I moved it to a AWS system with 512G ram (a x2iedn.4xlarge) and that's where it was able to run long enough to have this issue.  Is her code not great?  Almost certainly, but she's a virologist, not a software engineer ;) She's just trying to publish a paper to which this data will hopefully contribute, and will likely never use that particular set of code again (she'd use it as lessons-learned for any subsequent need, or pay a bioinformatics person to do it).  I think that's a very frequent use case for R and seurat - someone just trying to get something done, where doubling the hardware resources is more viable than spending weeks writing it well.

Is it worth there being a quick seurat vignette on how to possibly get around the issue, as a stop-gap to a backend change to the object structure?
